packages/toolkit/src/setup.ts

```diff
@@ -0,0 +1,182 @@
+  | AragonNetworkTag
+  | DefaultNetworkTag
+  | 'aragen'
+  | string
+  | ethers.providers.Provider
+  | AsyncSendable
+
+export interface ProviderOptions {
+  ipfsGateway?: string
+  ensAddress?: string
+  infuraProjectId?: string
+  etherscanApiToken?: string
+  // Options for JsonRpc providers
+  headers?: { [key: string]: string | number }
+  timeout?: number
+}
+
+type DefaultNetworkTag =
+  | 'homestead'
+  | 'rinkeby'
+  | 'ropsten'
+  | 'kovan'
+  | 'goerli'
+const defaultNetworkTags: DefaultNetworkTag[] = [
+  'homestead',
+  'rinkeby',
+  'ropsten',
+  'kovan',
+  'goerli',
+]
+
+type AragonNetworkTag =
+  | 'aragon:mainnet'
+  | 'aragon:ropsten'
+  | 'aragon:rinkeby'
+  | 'aragon:kovan'
+const aragonNetworkTags: AragonNetworkTag[] = [
+  'aragon:mainnet',
+  'aragon:ropsten',
+  'aragon:rinkeby',
+  'aragon:kovan',
+]
+
+/**
+ * Returns an initialized ethers provider given a flexible options
+ *
+ * @param providerArg Provider argument
+ * - Aragon network name {AragonNetworkTag}
+ *   'aragon:mainnet', 'aragon:ropsten', 'aragon:rinkeby', 'aragon:kovan'
+ *   Connects to Aragon's public nodes
+ * - Default network name {DefaultNetworkTag}
+ *   'homestead', 'rinkeby', 'ropsten', 'kovan', 'goerli'
+ *   Connects to a public RPC endpoint (Infura, Etherscan)
+ *   If no token is provided the default ethers token is used.
+ *   You can provide your Infura project ID with the option `infuraProjectId`
+ *   or your own Etherscan api token with the option `etherscanApiToken`
+ * - 'aragen'
+ *   Connect to a local testnet with Aragon's aragen settings
+ * - JSON RPC URL endpoint {string}
+ *   Connects to the given valid URL
+ * - ethers provider {ethers.providers.Provider}
+ *   Already initialized ethers instance
+ * - web3 provider {AsyncSendable}
+ *   Web3.js provider. Must have the sendAsync method
+ *
+ * @param options Additional options
+ * - ipfsGateway: IPFS gateway to retrieve content if necessary
+ * - ensAddress: Custom ENS address for private networks
+ * - infuraProjectId: Infura project ID to prevent using the default one
+ * - etherscanApiToken?: Etherscan api token to prevent using the default one
+ * - headers: Custom headers for JSON RPC providers
+ * - timeout: Custom timeout for JSON RPC providers
+ */
+export function parseProviderArgument(
+  providerArg?: ProviderArgument,
+  options?: ProviderOptions
+): ethers.providers.Provider {
+  const { ensAddress, headers, timeout, infuraProjectId, etherscanApiToken } =
+    options || {}
+
+  /**
+   * Ethers network options allow to provide a custom ensAddress
+   * Most ethers providers accept it so it's defined in the top scope
+   */
+  const networkOptions = ensAddress
+    ? { name: 'custom', chainId: 999, ensAddress }
+    : undefined
+
+  function getProviderFromUrl(url: string): ethers.providers.Provider {
+    const connectionInfo = headers ? { url, headers, timeout } : url
+    return new ethers.providers.JsonRpcProvider(connectionInfo, networkOptions)
```

Comment:
We might want to use a JsonRpcWebSocketProvider now that's supported by ethers: https://github.com/ethers-io/ethers.js/issues/141

Reply:
But it's on the v5 beta tho